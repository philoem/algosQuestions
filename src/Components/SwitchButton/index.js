import React from 'react'
import PropTypes from 'prop-types'
import styles from './index.module.css'

const SwitchButton = ({ onClick }) => {
	return (
		<div className={styles.containerSwitchBtn}>
			<label className={styles.switch}>
				<input aria-label='checkbox' type='checkbox' onClick={onClick} />
				<span className={styles.slider} />
			</label>
		</div>
	)
}

SwitchButton.propTypes = {
	onClick: PropTypes.func
}

SwitchButton.defaultProps = {
	onClick: () => null
}

export default SwitchButton
