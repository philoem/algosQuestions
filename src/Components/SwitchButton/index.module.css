.containerSwitchBtn {
  display: flex;
  flex-direction: row;
  align-items: flex-end;
  box-sizing: border-box;
  height: 12px;
  padding-left: 10px;
}

.switch input
{
  display: none;
}

.switch 
{
  display: inline-block;
  width: 60px; /*=w*/
  height: 20px; /*=h*/
  margin: 4px;
  transform: translateY(140%);
  position: relative;
}

.slider
{
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  border-radius: 30px;
  box-shadow: 0 0 0 2px #777, 0 0 4px #777;
  cursor: pointer;
  border: 4px solid transparent;
  overflow: hidden;
  transition: 0.2s;
}

.slider:before
{
  position: absolute;
  content: "";
  width: 100%;
  height: 100%;
  background-color: #777;
  border-radius: 30px;
  transform: translateX(-65px); /*translateX(-(w-h))*/
  transition: 0.2s;
}

input:checked + .slider:before
{
  transform: translateX(13px); /*translateX(w-h)*/
  background-color: limeGreen;
}

input:checked + .slider
{
  box-shadow: 0 0 0 2px limeGreen, 0 0 8px limeGreen;
}