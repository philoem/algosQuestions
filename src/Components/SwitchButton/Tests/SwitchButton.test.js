import React from 'react'
import { render, screen } from '@testing-library/react'
import { defineFeature, loadFeature } from 'jest-cucumber'
import SwitchButton from '../../SwitchButton/index'
import userEvent from '@testing-library/user-event'

const feature = loadFeature('src/Components/SwitchButton/Tests/SwitchButton.feature')

defineFeature(feature, (test) => {
	let container
	const onclick = jest.fn()

	const givenSwitchButtonIsRendered = (given, comp) => {
		given('Switch button is rendered', () => {
			container = render(comp).container
		})
	}

	const andRequiredElementsAreMounted = (and) => {
		and('Required elements are mounted', () => {
			expect(screen.getByRole('checkbox')).toBeInTheDocument()
		})
	}

	const andDefaultStyleIsApplied = (and) => {
		and('Default style is applied', () => {
			expect(container).toMatchSnapshot()
		})
	}
	test('Switch button clicked for the first time', ({ given, when, then, and }) => {
		givenSwitchButtonIsRendered(given, <SwitchButton onClick={onclick} />)
		andRequiredElementsAreMounted(and)
		andDefaultStyleIsApplied(and)

		when('Switch button is clicked', () => {
			userEvent.click(screen.getByRole('checkbox'))
		})

		then('Switch button is checked', () => {
			const checkbox = screen.getByRole('checkbox', { name: 'checkbox' })
			expect(checkbox).toBeChecked()
		})

		and('Switch button callback should be called', () => {
			expect(onclick).toHaveBeenCalledTimes(1)
		})
	})
})
