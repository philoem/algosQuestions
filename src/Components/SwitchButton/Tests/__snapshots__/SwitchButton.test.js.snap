// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Button clickable Switch button clicked for the first time 1`] = `
<div>
  <div
    class="containerSwitchBtn"
  >
    <label
      class="switch"
    >
      <input
        aria-label="checkbox"
        type="checkbox"
      />
      <span
        class="slider"
      />
    </label>
  </div>
</div>
`;
